### PR TITLE
add zoo model

### PR DESCRIPTION
**Pull request by simanw304**

### CODE REVIEW

File: examples/wikiqa/config/knrm_wikiqa.config [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "net_name": "KNRM",
   "global":{
       "model_type": "PY",
-      "weights_file": "examples/wikiqa/weights/knrm.wikiqa.weights",
+      "weights_file": "./examples/wikiqa/weights/knrm.wikiqa.weights",
```
Comment (zhichao-li) [inline examples/wikiqa/config/knrm_wikiqa.config]: `./` can be removed as well.

File: matchzoo/losses/zrank_losses.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    return rank_loss.__name__
+
+
+def deserialize(name, custom_objects=None):
+    return deserialize_keras_object(name,
+                                    module_objects=globals(),
+                                    custom_objects=custom_objects,
+                                    printable_module_name='loss function')
+
+def rank_hinge_loss(kwargs=None):
```
Comment (zhichao-li) [inline matchzoo/losses/zrank_losses.py]: Wondering would it be possible if we only simply add a new method `z_rank_hinge_loss`  in `rank_losses`

Comment (zhichao-li) [reply matchzoo/losses/zrank_losses.py]: is this useless as you've defined it in `zmain.py` ?

File: matchzoo/zmain.py [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+        mobj = mobj.lower()
+        if '@' in mobj:
+            mt_key, mt_val = mobj.split('@', 1)
+            eval_metrics[mobj] = metrics.get(mt_key)(int(mt_val))
+        else:
+            eval_metrics[mobj] = metrics.get(mobj)
+
+    zmodel.compile(optimizer='adam', loss=zloss(batch=10))
+    print('[Model] Model Compile Done.', end='\n')
+    zoo_input_data = []
```
Comment (zhichao-li) [inline matchzoo/zmain.py]: let's wrap this into a method: `gen_zinput`

File: matchzoo/zmain.py [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+                        res[k] += eval_func(y_true = y_true, y_pred = y_pred)
+                    num_valid += 1
+            generator.reset()
+            print('Iter:%d\t%s' % (i_e, '\t'.join(['%s=%f'%(k,v/num_valid) for k, v in res.items()])), end='\n')
+            sys.stdout.flush()
+        # if (i_e+1) % save_weights_iters == 0:
+        #     zmodel.save_weights(weights_file % (i_e+1))
+
+
+def standardize_input_data(data, names, shapes=None,
```
Comment (zhichao-li) [inline matchzoo/zmain.py]: try import this method like: `from keras.engine.training import _standardize_input_data`

**Comment by simanw304** [comment]
sgd
loss is 0.18508473412033968
loss is 0.17911778150924615
loss is 0.16186983125018223
loss is 0.21198988820625736
[07-23-2018 16:45:09]	[Eval:test] Iter:0	ndcg@3=0.593858	map=0.597403	ndcg@5=0.638770
[07-23-2018 16:45:11]	[Eval:valid] Iter:0	ndcg@3=0.602069	map=0.605475	ndcg@5=0.659650

adam
loss is 0.0
[07-23-2018 17:07:10]	[Eval:test] Iter:0	ndcg@3=0.224309	map=0.293743	ndcg@5=0.318502
[07-23-2018 17:07:12]	[Eval:valid] Iter:0	ndcg@3=0.237758	map=0.291279	ndcg@5=0.316762

original model output
[07-23-2018 16:58:03]	[Train:train] Iter:397	loss=0.000160
[07-23-2018 16:58:04]	[Eval:test] Iter:397	ndcg@3=0.478643	map=0.502947	ndcg@5=0.538765
[07-23-2018 16:58:04]	[Eval:valid] Iter:397	ndcg@3=0.568109	map=0.554744	ndcg@5=0.613828
[07-23-2018 16:58:04]	[Train:train] Iter:398	loss=0.000338
[07-23-2018 16:58:05]	[Eval:test] Iter:398	ndcg@3=0.481544	map=0.504570	ndcg@5=0.541978
[07-23-2018 16:58:06]	[Eval:valid] Iter:398	ndcg@3=0.560986	map=0.551503	ndcg@5=0.622051
[07-23-2018 16:58:06]	[Train:train] Iter:399	loss=0.000445
[07-23-2018 16:58:07]	[Eval:test] Iter:399	ndcg@3=0.483201	map=0.506756	ndcg@5=0.546033
[07-23-2018 16:58:07]	[Eval:valid] Iter:399	ndcg@3=0.560224	map=0.551025	ndcg@5=0.621528
